src/compiler/checker.ts

```diff
@@ -8040,7 +8076,24 @@ namespace ts {
             if (needToCaptureLexicalThis) {
                 captureLexicalThis(node, container);
             }
-
+            if (isFunctionLike(container)) {
+                const type = getContextuallyTypedThisType(container);
+                if (type) {
+                    return type;
+                }
+                const signature = getSignatureFromDeclaration(container);
+                if (signature.thisType) {
+                    return signature.thisType;
+                }
+                if (container.parent && container.parent.kind === SyntaxKind.ObjectLiteralExpression) {
+                    // Note: this works because object literal methods are deferred,
+                    // which means that the type of the containing object literal is already known.
+                    const type = checkExpressionCached(<ObjectLiteralExpression>container.parent);

```

Comment:
Huh - we should add a regression test to ensure that a set of mutually recursive object literal methods type-checks properly.

```ts
var x = {
    a: 100,

    start() {
        return this.doX(this.a);
    }    

    doX(n: number) {
        this.doY(n);
    }

    doY(n: number): number {
        if (n > 10) {
            return this.doX(n - 10);
        }
        if (n > 5) {
            return this.doY(n - 1);
        }
        return n;
    }
}

var i: number = x.start();
```

And maybe create an interface with the same shape, create a variable of that type, and assign `x` to it.